astropy/utils/data.py

```diff
@@ -1327,43 +1406,303 @@ def _get_download_cache_locs():
 
 # the cache directory must be locked before any writes are performed.  Same for
 # the hash shelve, so this should be used for both.
-def _acquire_download_cache_lock():
-    """
-    Uses the lock directory method.  This is good because `mkdir` is
-    atomic at the system call level, so it's thread-safe.
-    """
-
+#
+# In fact the shelve, if you're using gdbm (the default on Linux), it can't
+# be open for reading if someone has it open for writing. So we need to lock
+# access to the shelve even for reading.
+@contextlib.contextmanager
+def _cache_lock(need_write=False):
     lockdir = os.path.join(_get_download_cache_locs()[0], 'lock')
-    for i in range(conf.download_cache_lock_attempts):
-        try:
-            os.mkdir(lockdir)
-            # write the pid of this process for informational purposes
-            with open(os.path.join(lockdir, 'pid'), 'w') as f:
-                f.write(str(os.getpid()))
+    pidfn = os.path.join(lockdir, 'pid')
+    assume_cache_readonly = False
+    try:
+        msg = ("Config file requests {} tries"
+               .format(conf.download_cache_lock_attempts))
+        waited = 0.
+        while True:
+            try:
+                os.mkdir(lockdir)
+            except FileExistsError:
+                # It is not safe to open and inspect the pid file here
+                # on Windows, because while it is open that prevents its
+                # deletion and that of the directory containing it. This
+                # gets in the way of useful error messages.
+                msg = (
+                    "Cache is locked after {:.2f} s. This may indicate an "
+                    "astropy bug or that kill -9 "
+                    "was used. If you want to unlock the cache remove the "
+                    "directory or file \"{}\"."
+                    .format(waited, lockdir)
+                )
+            except PermissionError:
+                assume_cache_readonly = True
+                if need_write:
+                    raise
+                else:
+                    break
+            else:
+                # Got the lock!
+                # write the pid of this process for informational purposes
+                with open(pidfn, 'w') as f:
+                    f.write(str(os.getpid()))
+                break
+            if waited < conf.download_cache_lock_attempts:
```

Comment:
It's worse than that: it's not a number of tries at all any more, it's a number of seconds to wait. I didn't want to change the name because it would exist in people's config files. Now, there's not much reason to ever change this, and users who have increased the number of tries will probably find they don't need to increase this because the new code doesn't spend very long locked, but I wasn't sure whether it was okay to ignore an old config option and introduce a new one.

Reply:
I'm not sure which text you are recommending changing here?

Reply:
https://github.com/astropy/astropy/blob/b0a84f612ab6c058c60b3333668ef21878df7c3f/astropy/utils/data.py#L65

Basically changing "number of tries" to "number of seconds," if I understood you correctly.

Reply:
I have changed this text, I think we should be okay. I did wonder about the "aliases" business, which probably exists for just this sort of purpose.

Reply:
The "aliases" is (are?) for when we move the config around. For example, when we move `astropy.sub.subpackage` to `astropy.subpackage`, so that `astropy.sub.subpackage.conf` is now `astropy.subpackage.conf`. It also can be (ab?)used to deprecate a config item in the same subpackage.

Reply:
To clarify, you do not need to use "aliases" for this PR.